File: back/utentes/lib/validator.py
# -*- coding: utf-8 -*-

import datetime
import dateutil.parser

class IsNotNull():

    def fails(self, value):
        if value is None:
            return True
        return False


class IsDate():
    '''
    The received value is a valid datetime.date object or can be parsed with
    dateutil.parser. This allows strings in ISO8601 or RFC3329 or others
    '''
    def fails(self, value):
        if not value:
            return False
        # datetime.datetime is child of datetime.date
        if instanceof(value, datetime.date):
            return False
        try:
            dateutil.parser.parse(isohoy)
        except:
            return True

        return False

class IsNumeric():
    '''
    The received value is the representation of a number. So '5' is considered valid
    '''
    def fails(self, value):
        if not value:
            return False
        try:
            float(value)
        except:
            return True
        return False


class Validator():

    def __init__(self, schemaValidateFrom):
        self.messages = []
        self.schema = schemaValidateFrom
        self.rules = {
            'NOT_NULL': IsNotNull(),
            'IS_DATE':  IsDate(),
            'IS_NUMERIC': IsNumeric()
        }

    def validate(self, model):
        self.messages = []

        for definition in self.schema:
            for rulename in definition['rules']:
                rule = self.get_rule(rulename)
                print rule
                if isinstance(rule, dict):
                    print model
                    model[definition['fieldname']]
                    if rule['fails'](model[definition['fieldname']]):
                        self.messages.append(definition['message'])
                else:
                    if rule.fails(model[definition['fieldname']]):
                        self.messages.append(definition['message'])

        return self.messages

    def get_rule(self, rulename):
        return self.rules[rulename]

    def add_rule(self, rulename, rule_def):
        self.rules[rulename] = rule_def

File: back/utentes/static/lib/schema-validator/validator.js
function validator(schemaValidateFrom) {
    var validatorObj = new Object();

    var rules = {
        NOT_NULL: isNotNull(),
        IS_DATE: isDate(),
        IS_NUMERIC: isNumeric(),
        INT_LESS_THAN_8: isIntLessThan8(),
        IS_BOOLEAN: isBoolean(),
        ARRAY_NOT_VOID: isArrayNotVoid(),
    };

    var messages = [];
    var schema = schemaValidateFrom;

    function validate(model) {
        messages = [];
        messages.length = 0;

        schema.forEach(function(def) {
            // if any of the functions defined in the context key is true apply the rules
            if (
                !def.context?.some(context =>
                    SIRHA.Services.ValidatorContextService[context](model)
                )
            ) {
                return messages;
            }

            def.rules.forEach(function(ruleName) {
                var rule = getRule(ruleName);
                if (rule.fails(model[def.fieldname])) {
                    messages.push(def.message);
                }
            });
        });

        return messages;
    }

    function getRule(ruleName) {
        return rules[ruleName];
    }

    function addRule(ruleName, ruleDef) {
        rules[ruleName] = ruleDef;
    }

    function appendSchema(aSchema) {
        aSchema.forEach(function(e) {
            schema.push(e);
        });
    }

    function isNotNull() {
        var ruleObj = new Object();
        function fails(value) {
            if (value === "" || value === undefined || value === null) {
                return true;
            }
            return false;
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    function isDate() {
        var ruleObj = new Object();
        var isoFormat = /^\d{4}-\d{2}-\d{2}$/;

        function fails(value) {
            var valid =
                value instanceof Date || value === null || isoFormat.test(value);
            return !valid;
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    // check if value is the representation of a number. So '5'
    // is considered valid
    function isNumeric() {
        var ruleObj = new Object();

        function fails(value) {
            return value && !(!isNaN(parseFloat(value)) && isFinite(value));
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    function isIntLessThan8() {
        var ruleObj = new Object();

        function fails(value) {
            if (value && value > 99999999.99) {
                return true;
            }
            return false;
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    function isBoolean() {
        var ruleObj = new Object();

        function fails(value) {
            return value != true && value != false && value != null;
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    function isArrayNotVoid() {
        var ruleObj = new Object();

        function fails(value) {
            return !Array.isArray(value) || value.length === 0;
        }

        ruleObj.fails = fails;
        return ruleObj;
    }

    validatorObj.validate = validate;
    validatorObj.getRule = getRule;
    validatorObj.addRule = addRule;
    validatorObj.appendSchema = appendSchema;
    return validatorObj;
}
